packages/protocol/contracts/ACE/validators/adjustSupply/AdjustSupply.sol

```diff
@@ -0,0 +1,384 @@
+ * External calls to OptimizedAZTEC can be treated as pure functions as this contract contains no
+ * storage and makes no external calls (other than to precompiles)
+ * Copyright Spilsbury Holdings Ltd 2019. All rights reserved.
+ * We will be releasing AZTEC as an open-source protocol that provides efficient transaction privacy for Ethereum.
+ * Our full vision of the protocol includes confidential cross-asset interactions via our
+ * family of AZTEC zero-knowledge proofs
+ * and the AZTEC token standard, stay tuned for updates!
+ **/
+contract AdjustSupply is LibEIP712 {
+    /**
+     * @dev AZTEC will take any transaction sent to it and attempt to validate a zero knowledge proof.
+     * If the proof is not valid, the transaction will throw.
+     * @notice See AZTECInterface for how method calls should be constructed.
+     * 'Cost' of raw elliptic curve primitives for a transaction:
+     * 260,700 gas + (124,500 * number of input notes) + (167,600 * number of output notes).
+     * For a basic 'AdjustSupply' with 2 inputs and 2 outputs = 844,900 gas.
+     * AZTEC is written in YUL to enable manual memory management and for other efficiency savings.
+     **/
+    function() external payable {
+        assembly {
+            // We don't check for function signatures,
+            // there's only one function that ever gets called: validateAdjustSupply()
+            // We still assume calldata is offset by 4 bytes so that we can represent this contract
+            // through a compatible ABI
+            validateAdjustSupply()
+
+            /**
+             * New calldata map
+             * 0x04:0x24      = calldata location of proofData byte array
+             * 0x24:0x44      = message sender
+             * 0x44:0x64      = h_x
+             * 0x64:0x84      = h_y
+             * 0x84:0xa4      = t2_x0
+             * 0xa4:0xc4      = t2_x1
+             * 0xc4:0xe4      = t2_y0
+             * 0xe4:0x104     = t2_y1
+             * 0x104:0x124    = length of proofData byte array
+             * 0x124:0x144    = challenge
+             * 0x144:0x164    = offset in byte array to notes
+             * 0x1a4:0x1c4    = offset in byte array to inputOwners
+             * 0x1c4:0x1e4    = offset in byte array to outputOwners
+             * 0x1e4:0x204    = offset in byte array to metadata
+             */
+            function validateAdjustSupply() {
+                mstore(0x80, calldataload(0x44))
+                mstore(0xa0, calldataload(0x64))
+                let notes := add(0x104, calldataload(0x144)) // get the length of notes
+                let m := 1
+                let n := calldataload(notes)
+                let gen_order := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
+                let challenge := mod(calldataload(0x124), gen_order)
+
+
+                // validate m <= n
+                if gt(m, n) { mstore(0x00, 404) revert(0x00, 0x20) }
```

Comment:
we need to check that (m < n) - we need at least 1 output note as well as an input note

Reply:
Done